index.js

```diff
@@ -1,8 +1,90 @@
-const net = require('net');
+'use strict';
+
+const fs = require('fs');
+const http = require('http');
+const path = require('path');
+
+const constants = require('./src/common/constants');
 const server = require('./src/server');
 
+const tempDir = path.join(__dirname, 'tmp')
+const statusFilePath = path.join(tempDir, 'status.txt')
+
 module.exports = {
-    startServer: (port) => {
-        return server.start(port || 0);
+    startServer: () => {
+        let currentServerPort;
+        return getState()
+            .then(port => checkServerHealth(port)
+                .then(serverHealth => {
+                    currentServerPort = port
+                    if (!serverHealth) {
+                        return server.start(port)
+                            .catch(initServerWithFreePort)
+                            .then(port => {
+                                currentServerPort = port;
+                                watchStatusFile();
+                                return currentServerPort;
+                            })
+                    }
+
+                    return currentServerPort;
+                })
+            , err => initServerWithFreePort()
+                .then(port => {
+                    currentServerPort = port;
+                    watchStatusFile();
+                    return currentServerPort;
+                })
+            )
+    }
+}
+
+function initServerWithFreePort() {
+    return server.start(0)
+        .then(saveState);
+}
+
+function saveState(port) {
+    return new Promise((resolve, reject) => {
+        ensureDirExists(tempDir);
+        fs.writeFile(statusFilePath, port, err => err ? reject(err) : resolve(port))
+    });
+}
+
+function getState() {
+    return new Promise((resolve, reject) => {
+        ensureDirExists(tempDir);
```

Comment:
Semantically you needn't ensure the `tempDir` existence here - if it doesn't exist then neither will the `statusFilePath`